pkg/operator/ceph/cluster/watcher.go

```diff
@@ -53,61 +56,61 @@ func (c *clientCluster) onK8sNode(object runtime.Object) bool {
 	if !ok {
 		return false
 	}
+	// Check if the node have OSDs already present on it, reconcile if OSDs are not present
+	if !nodesCheckedForReconcile.Contains(node.Name) {
+		// Get CephCluster
+		cluster := c.getCephCluster()
 
-	// Get CephCluster
-	cluster := c.getCephCluster()
-
-	if !k8sutil.GetNodeSchedulable(*node) {
-		logger.Debugf("node watcher: skipping cluster update. added node %q is unschedulable", node.Labels[v1.LabelHostname])
-		return false
-	}
-
-	if !k8sutil.NodeIsTolerable(*node, cephv1.GetOSDPlacement(cluster.Spec.Placement).Tolerations, false) {
-		logger.Debugf("node watcher: node %q is not tolerable for cluster %q, skipping", node.Name, cluster.Namespace)
-		return false
-	}
-
-	if !cluster.Spec.Storage.UseAllNodes {
-		logger.Debugf("node watcher: do not use all Nodes in cluster %q, skipping", cluster.Namespace)
-		return false
-	}
+		if !k8sutil.GetNodeSchedulable(*node) {
+			logger.Debugf("node watcher: skipping cluster update. added node %q is unschedulable", node.Labels[v1.LabelHostname])
+			return false
+		}
 
-	// Too strict? this replaces clusterInfo == nil
-	if cluster.Status.Phase != cephv1.ConditionReady {
-		logger.Debugf("node watcher: cluster %q is not ready. skipping orchestration", cluster.Namespace)
-		return false
-	}
+		if !k8sutil.NodeIsTolerable(*node, cephv1.GetOSDPlacement(cluster.Spec.Placement).Tolerations, false) {
+			logger.Debugf("node watcher: node %q is not tolerable for cluster %q, skipping", node.Name, cluster.Namespace)
+			return false
+		}
 
-	valid, _ := k8sutil.ValidNode(*node, cephv1.GetOSDPlacement(cluster.Spec.Placement))
-	if valid {
-		nodeName := node.Name
-		hostname, ok := node.Labels[v1.LabelHostname]
-		if ok && hostname != "" {
-			nodeName = hostname
+		if !cluster.Spec.Storage.UseAllNodes {
+			logger.Debugf("node watcher: do not use all Nodes in cluster %q, skipping", cluster.Namespace)
+			return false
 		}
 
-		// Make sure we can call Ceph properly
-		// Is the node in the CRUSH map already?
-		// If so we don't need to reconcile, this is done to avoid double reconcile on operator restart
-		// Assume the admin key since we are watching for node status to create OSDs
-		clusterInfo := cephclient.AdminClusterInfo(cluster.Namespace)
-		osds, err := cephclient.GetOSDOnHost(c.context, clusterInfo, nodeName)
-		if err != nil {
-			// If it fails, this might be due to the the operator just starting and catching an add event for that node
-			logger.Debugf("failed to get osds on node %q", nodeName)
+		// Too strict? this replaces clusterInfo == nil
+		if cluster.Status.Phase != cephv1.ConditionReady {
+			logger.Debugf("node watcher: cluster %q is not ready. skipping orchestration", cluster.Namespace)
+			nodesCheckedForReconcile.Add(node.Name) // adding node to nodesCheckedForReconcile Set
```

Comment:
When a node comes online, we will get several different events. Initially it will not be ready, then we will get a new event when it is ready. We only want to add the node to the list when the node is ready. So let's move this call to `Add()` a couple lines lower out of this `if` condition so that we only add the node when it is finally ready.